boxtree/tree_build.py

```diff
@@ -46,6 +46,207 @@ class MaxLevelsExceeded(RuntimeError):
+    box_centers = resized_array(tob.box_centers, nboxes_new)
+    box_children = resized_array(tob.box_children, nboxes_new)
+    box_levels = resized_array(tob.box_levels, nboxes_new)
+
+    # new boxes are appended at the end, so coarsen_flags are still meaningful
+    box_parents[tob.nboxes:] = refine_parents_per_child
+    box_levels[tob.nboxes:] = tob.box_levels[box_parents[tob.nboxes:]] + 1
+    box_children[:, refine_parents] = (
+        child_box_starts
+        + np.arange(nchildren).reshape(-1, 1))
+
+    for i in range(2**tob.dim):
+        children_i = box_children[i, refine_parents]
+        offsets = (
+                tob.root_box_extent
+                * vec_of_signs(tob.dim, i).reshape(-1, 1)
+                * (1/2**(1+box_levels[children_i])))
+        box_centers[:, children_i] = (
+                box_centers[:, refine_parents]
+                + offsets)
+
+    # ------------------- coarsen only if all peers are leaves after refinement
+
+    coarsen_flags = resized_array(coarsen_flags, nboxes_new)
+    coarsen_sources, = np.where(coarsen_flags)
+    if coarsen_sources:
+        # FIXME
+        raise NotImplementedError
+        coarsen_parents = tob.box_parents[coarsen_sources]
+        coarsen_peers = tob.box_children[:, coarsen_parents]
+        coarsen_peer_is_leaf = tob.get_leaf_flags()[coarsen_peers]
+        coarsen_exec_flags = np.all(coarsen_peer_is_leaf, axis=0)
+
+        coarsen_flags_ignored = (coarsen_exec_flags != coarsen_flags)
+        if np.any(coarsen_flags_ignored):
+            import warnings
+            warnings.warn(f"{np.sum(coarsen_flags_ignored)} out of "
+                          f"{np.sum(coarsen_flags)} coarsening flags ignored "
+                          "to prevent removing non-leaf boxes")
+
+        coarsen_parents = coarsen_parents[coarsen_exec_flags]
+        coarsen_peers = coarsen_peers[:, coarsen_exec_flags]
+        box_parents[coarsen_peers] = -1
+        box_children[coarsen_parents] = 0
+
+    return TreeOfBoxes(
+            box_centers=box_centers,
+            root_box_extent=tob.root_box_extent,
+            box_parents=box_parents,
+            box_children=box_children,
+            box_levels=box_levels)
+
+
+# }}} End refine/coarsen
+
+
+def make_tob_root(dim, bbox):
+    """
+    Make the minimal tree of boxes.
+    """
+    center = np.array(
+        [(bbox[0][iaxis] + bbox[1][iaxis]) * 0.5 for iaxis in range(dim)])
+    extents = np.array(
+        [(bbox[1][iaxis] - bbox[0][iaxis]) for iaxis in range(dim)])
+    from pytools import single_valued
+    return TreeOfBoxes(
+            box_centers=center.reshape(dim, 1),
+            root_box_extent=single_valued(extents, np.allclose),
+            box_parents=np.array([0], np.intp),
+            box_children=np.array([0] * 2**dim, np.intp).reshape(2**dim, 1),
+            box_levels=np.array([0]),
+            )
+
+
+def distribute_quadrature_rule(tob, quad_rule):
```

Comment:
I'm not sure I love this functionality. I think I'd prefer if the equivalent thing just appeared "out of thin air" via introducing the meshmode interop.